pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -137,23 +138,59 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, pus
-		if service == nil {
-			// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
-			// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
-			// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
-			for _, n := range watched.ResourceNames {
-				_, _, svcHost, _ := model.ParseSubsetKey(n)
-				if svcHost == host.Name(key.Name) {
-					removedClusterNames = append(removedClusterNames, n)
-				}
+		if key.Kind == gvk.ServiceEntry {
+			service := push.ServiceForHostname(proxy, host.Name(key.Name))
+			// SidecarScope.ServiceForHostname will return nil if the proxy doesn't care about the service OR it was deleted.
+			// we can cross-reference with WatchedResources to figure out which services were deleted.
+			if service == nil {
+				// we assume a service was deleted, it doesn't hurt to have positives for removedResources when a resource included
+				// does not exist.
+				// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
+				// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
+				// so, we can check if the cluster names contains the service name, and determine if it is deleted by that
+				removedClusterNames = append(removedClusterNames, getRemovedClusterNames(watched.ResourceNames, nil, key.Name)...)
+			} else {
+				services = append(services, service)
 			}
 		} else {
-			services = append(services, service)
+			// destination rules
+			cfg := proxy.SidecarScope.DestinationRuleByName(key.Name)
+			if cfg == nil {
+				// a destinationrule was deleted, get the populated destinationRule from PrevSidecarScope
+				prevCfg := proxy.PrevSidecarScope.DestinationRuleByName(key.Name)
+				if prevCfg == nil {
+					// something went wrong, skip -- we shouldn't be nil here
+					break
+				}
+				dr := prevCfg.Spec.(*networking.DestinationRule)
+				removedClusterNames = append(removedClusterNames, getRemovedClusterNames(watched.ResourceNames, nil, dr.Host)...)
+			} else {
+				// destination exists, was updated
+				// generate cluster based on service associated with destinationrule
+				// hostName can be a wildcard, and can match to multiple services
+				dr := cfg.Spec.(*networking.DestinationRule)
+				matchedSvcs := push.ServicesForHostname(proxy, host.Name(dr.Host))
+				// service := push.ServiceForHostname(proxy, host.Name(dr.Host))
+				services = append(services, matchedSvcs...)
+				// check for removed subsets
+				prevCfg := proxy.PrevSidecarScope.DestinationRuleByName(key.Name)
+				if prevCfg == nil {
+					break
+				}
+				prevDr := prevCfg.Spec.(*networking.DestinationRule)
+				// detecting added/updated subsets works, we just need to track deletions
+				if !reflect.DeepEqual(prevDr.GetSubsets(), dr.GetSubsets()) {
```

Comment:
checking len is sufficient? Do not have to do complete equals check?

Reply:
if a subset is removed but then a new one is added, and then configuration is applied,  length will be the same, right? I think we need deepequals

Reply:
Yeah. Let us leave it as DeepEquals.

Reply:
Can add a TODO to build a more optimized equals for subsets?